### PR TITLE
:card_file_box: migrate trips table to use ID instead of IBNR

### PR DESCRIPTION
**Pull request by MrKrisKrisu**
see #2411

### CODE REVIEW

**Comment by MrKrisKrisu** [comment]
"SQLite doesn't support dropping foreign keys (you would need to re-create the table)."

(╯°□°)╯︵ ┻━┻

File: database/migrations/2023_11_21_000002_add_default_permissions.php [hunk 1]
```diff
@@ -1,61 +1,10 @@
...
-        $admins = User::role('admin')->get();
-        foreach ($admins as $admin) {
-            $admin->update(['role' => 10]);
-        }
-
-        Role::where('name', 'admin')->delete();
-        Role::where('name', 'event-moderator')->delete();
-        Role::where('name', 'open-beta')->delete();
-        Role::where('name', 'closed-beta')->delete();
+        // empty now - replaced by PermissionSeeder
```
Comment (HerrLevin) [inline database/migrations/2023_11_21_000002_add_default_permissions.php]: Hm. Not a fan. Why is this needed?
Also:
- This file should be deleted
- Moving this logic from a migration to a seeder should be a separate PR (imho)

File: database/migrations/2023_12_17_000003_add_event_history_permission.php [hunk 2]
```diff
@@ -7,11 +7,6 @@
 return new class extends Migration
 {
     public function up(): void {
-        $permission = Permission::create(['name' => 'view event history']);
-        Role::findByName('admin')->givePermissionTo($permission);
-    }
-
-    public function down(): void {
-        Permission::findByName('view event history')->delete();
+        // empty now - replaced by PermissionSeeder
```
Comment (HerrLevin) [inline database/migrations/2023_12_17_000003_add_event_history_permission.php]: Hm. Not a fan. Why is this needed?
Also:
- This file should be deleted
- Moving this logic from a migration to a seeder should be a separate PR (imho)

File: database/migrations/2024_01_29_000000_add_station_permissions.php [hunk 3]
```diff
@@ -7,28 +7,6 @@
...
-
-        $roleAdmin = Role::findByName('admin');
-        $roleAdmin->revokePermissionTo($permissionCreate);
-        $roleAdmin->revokePermissionTo($permissionUpdate);
-        $roleAdmin->revokePermissionTo($permissionDelete);
-
-        $permissionCreate->delete();
-        $permissionUpdate->delete();
-        $permissionDelete->delete();
+        // empty now - replaced by PermissionSeeder
```
Comment (HerrLevin) [inline database/migrations/2024_01_29_000000_add_station_permissions.php]: Hm. Not a fan. Why is this needed?
Also:
- This file should be deleted
- Moving this logic from a migration to a seeder should be a separate PR (imho)

File: database/migrations/2024_01_30_000000_add_activity_permissions.php [hunk 4]
```diff
@@ -7,18 +7,6 @@
...
-    }
-
-    public function down(): void {
-        $permission = Permission::findByName('view activity');
-
-        $roleAdmin = Role::findByName('admin');
-        $roleAdmin->revokePermissionTo($permission);
-
-        $permission->delete();
+        // empty now - replaced by PermissionSeeder
```
Comment (HerrLevin) [inline database/migrations/2024_01_30_000000_add_activity_permissions.php]: Hm. Not a fan. Why is this needed?
Also:
- This file should be deleted
- Moving this logic from a migration to a seeder should be a separate PR (imho)

File: database/migrations/2024_02_01_000000_add_permissions_for_manual_trips_beta.php [hunk 5]
```diff
@@ -1,25 +1,11 @@
...
-    }
-
-    public function down(): void {
-        $permission = Permission::findByName('create-manual-trip');
-
-        Role::findByName('admin')->revokePermissionTo($permission);
-        Role::findByName('closed-beta')->revokePermissionTo($permission);
-
-        $permission->delete();
+        // empty now - replaced by PermissionSeeder
```
Comment (HerrLevin) [inline database/migrations/2024_02_01_000000_add_permissions_for_manual_trips_beta.php]: Hm. Not a fan. Why is this needed?
Also:
- This file should be deleted
- Moving this logic from a migration to a seeder should be a separate PR (imho)

File: database/seeders/DatabaseSeeder.php [hunk 6]
```diff
@@ -3,14 +3,19 @@
...
+use App\Models\Trip;
+use Database\Seeders\Constants\PermissionSeeder;
 use Illuminate\Database\Seeder;
 
 class DatabaseSeeder extends Seeder
 {
 
     public function run(): void {
+        //Seed obligatory constants
+        $this->call(PermissionSeeder::class);
```
Comment (HerrLevin) [inline database/seeders/DatabaseSeeder.php]: Idk, I don't really like that.

File: app/Http/Controllers/API/v1/StationController.php [hunk 7]
```diff
@@ -99,8 +99,8 @@ public function merge(int $oldStationId, int $newStationId): StationResource {
         Checkin::where('origin', $oldStation->ibnr)->update(['origin' => $newStation->ibnr]);
         Checkin::where('destination', $oldStation->ibnr)->update(['destination' => $newStation->ibnr]);
         Stopover::where('train_station_id', $oldStation->id)->update(['train_station_id' => $newStation->id]);
-        Trip::where('origin', $oldStation->ibnr)->update(['origin' => $newStation->ibnr]);
```
Comment (HerrLevin) [inline app/Http/Controllers/API/v1/StationController.php]: Shouldn't you keep these for legacy Trips  without a origin/destination id?

File: app/Http/Controllers/HafasController.php [hunk 8]
```diff
@@ -419,8 +419,8 @@ public static function fetchHafasTrip(string $tripID, string $lineName): Trip {
                                          'linename'       => $tripJson->line->name,
                                          'journey_number' => $tripJson->line?->fahrtNr === "0" ? null : $tripJson->line?->fahrtNr,
                                          'operator_id'    => $operator?->id,
-                                         'origin'         => $origin->ibnr,
```
Comment (HerrLevin) [inline app/Http/Controllers/HafasController.php]: Shouldn't you keep these for legacy Trips  without a origin/destination id?

File: app/Models/Trip.php [hunk 9]
```diff
@@ -51,8 +50,8 @@ class Trip extends Model
         'category'       => HafasTravelType::class,
         'journey_number' => 'integer',
         'operator_id'    => 'integer',
-        'origin'         => 'integer',
```
Comment (HerrLevin) [inline app/Models/Trip.php:57]: Shouldn't you keep these for legacy Trips  without a origin/destination id?

File: app/Models/Trip.php [hunk 10]
```diff
@@ -66,11 +65,11 @@ public function polyline(): HasOne {
     }
 
     public function originStation(): BelongsTo {
-        return $this->belongsTo(Station::class, 'origin', 'ibnr');
```
Comment (HerrLevin) [inline app/Models/Trip.php]: Shouldn't you keep these for legacy Trips  without a origin/destination id?

File: app/Models/Trip.php [hunk 11]
```diff
@@ -66,11 +65,11 @@ public function polyline(): HasOne {
...
+        return $this->belongsTo(Station::class, 'origin_id', 'id');
     }
 
     public function destinationStation(): BelongsTo {
-        return $this->belongsTo(Station::class, 'destination', 'ibnr');
```
Comment (HerrLevin) [inline app/Models/Trip.php]: Shouldn't you keep these for legacy Trips  without a origin/destination id?

File: database/migrations/2024_03_11_000000_add_origin_and_destination_id_to_trips.php [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+        Schema::table('hafas_trips', static function(Blueprint $table) {
+            //Columns need to be nullable for the migration
+            $table->unsignedBigInteger('origin_id')->nullable()->after('origin');
+            $table->unsignedBigInteger('destination_id')->nullable()->after('destination');
+
+            $table->foreign('origin_id')->references('id')->on('train_stations');
+            $table->foreign('destination_id')->references('id')->on('train_stations');
+        });
+
+        DB::table('hafas_trips')->update([
```
Comment (HerrLevin) [inline database/migrations/2024_03_11_000000_add_origin_and_destination_id_to_trips.php:24]: Won't this be like... suuuuper slow? We currently have over 2 million trips. These 4 lines should equate to at least ~6 Million Queries.

**Review by HerrLevin — CHANGES_REQUESTED** [review]

File: database/migrations/2024_03_11_000000_add_origin_and_destination_id_to_trips.php [hunk 13]
```diff
@@ -0,0 +1,88 @@
...
+use Illuminate\Support\Facades\DB;
+use Illuminate\Support\Facades\Schema;
+
+/**
+ * "SQLite doesn't support multiple calls to dropColumn / renameColumn in a single modification."
+ * -> so small steps here. (╯°□°)╯︵ ┻━┻
+ */
+return new class extends Migration
+{
+    public function up(): void {
```
Comment (HerrLevin) [inline database/migrations/2024_03_11_000000_add_origin_and_destination_id_to_trips.php:14]: This migration should be split into multiple migrations and maybe moved into transactions.

**Comment by MrKrisKrisu** [comment]
@HerrLevin As discussed: I'll split up this PR in multiple small PR's and do separate migrations for create and drop the data for security reasons (data loss).

**Comment by HerrLevin** [comment]
@MrKrisKrisu Not necessarily multiple PRs but multiple migrations. :)

**Pull request closed without merging**